File: cakk-external/src/main/java/com/cakk/external/template/CertificationTemplate.java
package com.cakk.external.template;

import lombok.RequiredArgsConstructor;

import com.cakk.external.api.CertificationApiExecutor;
import com.cakk.external.api.CertificationMessageExtractor;
import com.cakk.external.vo.CertificationMessage;

@RequiredArgsConstructor
public class CertificationTemplate {

	private final CertificationApiExecutor certificationApiExecutor;
	private final CertificationMessageExtractor certificationMessageExtractor;

	public void sendMessageForCertification(CertificationMessage certificationMessage) {
		this.sendMessageForCertification(certificationMessage, certificationMessageExtractor, certificationApiExecutor);
	}

	public <T> void sendMessageForCertification(
		CertificationMessage certificationMessage,
		CertificationMessageExtractor certificationMessageExtractor,
		CertificationApiExecutor certificationApiExecutor
	) {
		T extractMessage = (T)certificationMessageExtractor.extract(certificationMessage);
		certificationApiExecutor.send(extractMessage);
	}
}

File: cakk-external/src/main/java/com/cakk/external/template/CertificationTemplate.kt
package com.cakk.external.template

import com.cakk.external.extractor.CertificationMessageExtractor
import com.cakk.external.sender.MessageSender
import com.cakk.external.vo.CertificationMessage

class CertificationTemplate<T>(
	private val messageSender: MessageSender<T>,
	private val certificationMessageExtractor: CertificationMessageExtractor<T>
) {

	fun sendMessageForCertification(certificationMessage: CertificationMessage) {
		this.sendMessageForCertification(
			certificationMessage = certificationMessage,
			certificationMessageExtractor = certificationMessageExtractor,
			messageSender = messageSender
		)
	}

	private fun <T> sendMessageForCertification(
		certificationMessage: CertificationMessage,
		certificationMessageExtractor: CertificationMessageExtractor<T>,
		messageSender: MessageSender<T>
	) {
		val extractMessage: T = certificationMessageExtractor.extract(certificationMessage)
		messageSender.send(extractMessage)
	}
}
